### PR TITLE
GG-24427 Save authenticated Webconsole session for more than one page refresh

### PR DESCRIPTION
**Pull request by nva**

### CODE REVIEW

File: modules/web-console/frontend/app/components/connected-clusters-dialog/components/list/column-defs.ts [hunk 1]
```diff
@@ -48,7 +48,7 @@ export const columnDefsFn = ($translate: ng.translate.ITranslateService) => [
                     value='COL_FIELD'
                 ></connected-clusters-cell-status>
                 <connected-clusters-cell-logout
-                    ng-if='row.entity.secured && grid.appScope.$ctrl.agentMgr.hasCredentials(row.entity.id)'
+                    ng-if='row.entity.secured'
```
Comment (akuznetsov-gridgain) [inline modules/web-console/frontend/app/components/connected-clusters-dialog/components/list/column-defs.ts]: Looks like that this `ng-if` may not work as expected after this change.

File: modules/web-console/frontend/app/modules/agent/types/ClusterSecrets.js [hunk 2]
```diff
@@ -35,24 +32,13 @@ export class ClusterSecrets {
...
         this.password = null;
     }
 
-    resetSessionToken() {
-        this.sessionToken = null;
-    }
-
     /**
-     * @return {{sessionToken: String}|{'user': String, 'password': String}}
+     * @return {{user: string, password: string}}
```
Comment (akuznetsov-gridgain) [inline modules/web-console/frontend/app/modules/agent/types/ClusterSecrets.js]: Do we need password at all?
For me we need flag "authenticated: boolean" instead.

File: modules/web-console/web-console-common/src/main/java/org/apache/ignite/console/websocket/WebSocketEvents.java [hunk 3]
```diff
@@ -58,4 +58,7 @@
 
     /** */
     public static final String CLUSTER_TOPOLOGY = "cluster:topology";
+
+    /** */
+    public static final String CLUSTER_LOGOUT = "cluster:logout";
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-console-common/src/main/java/org/apache/ignite/console/websocket/WebSocketEvents.java]: I think, after this change we need to increase web agent protocol version, should we?

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/repositories/IgniteSessionRepository.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+import org.springframework.session.SessionRepository;
+import org.springframework.stereotype.Repository;
+
+/**
+ * A {@link SessionRepository} backed by a Apache Ignite and that uses a {@link MapSession}.
+ */
+@Repository
+public class IgniteSessionRepository implements SessionRepository<ExpiringSession> {
+    /** The number of seconds that the {@link Session} should be kept alive between requests (default: 30 days). */
+    @Value("${sessions.inactive.timeout:2592000}")
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/repositories/IgniteSessionRepository.java]: `sessions.inactive.timeout` should be documented in README file.

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/BrowsersService.java [hunk 5]
```diff
@@ -319,54 +385,80 @@ protected void registerVisorTasks() {
...
+                }
+
+                res.clearSessionToken();
+
+                sendMessageQuiet(ses, evt0.response(res));
+            }
+            catch (Exception ignored) {
+                sendMessageQuiet(ses, evt);
+            }
+        } else if (ses != null)
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/BrowsersService.java]: `else` should be on new line

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/BrowsersService.java [hunk 6]
```diff
@@ -158,19 +188,55 @@ public void sendToAgent(AgentKey key, WebSocketRequest evt) {
...
+                        throw new IllegalStateException(messages.getMessageWithArgs("err.cluster-not-found-by-id", clusterId));
+
+                    // TODO GG-24427 Append sessionId if it exists in session store.
+                    if (top.isSecured()) {
+                        SessionAttribute sesAttr = new SessionAttribute(ses, SESSION_TOKEN_ATTR_PREFIX + clusterId);
+
+                        String sesTok = sesSrvc.get(sesAttr);
+
+                        if (F.isEmpty(sesTok) && Stream.of("user", "password").noneMatch(params::containsKey)) {
+                            sendMessageQuiet(ses, evt.response(RestResult.fail(STATUS_FAILED, "Failed to handle request - unknown session token (maybe expired session)")));
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/BrowsersService.java]: Message "Failed to handle request..." should be in resources.

File: modules/web-console/web-console-server/src/test/java/org/apache/ignite/console/services/SessionsServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    public static void tearDown() throws Exception {
+        stopAllGrids();
+        cleanPersistenceDir();
+    }
+
+    /**
+     * Should change session attribute.
+     */
+    @Test
+    public void interactionWithAttribute() {
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-console-server/src/test/java/org/apache/ignite/console/services/SessionsServiceTest.java]: can we have a test for "logout from cluster" scenario?

File: modules/web-console/frontend/app/modules/agent/AgentManager.service.ts [hunk 8]
```diff
@@ -373,8 +361,7 @@ export default class AgentManager {
         const oldCluster = _.find(conn.clusters, (cluster) => cluster.id === newCluster.id);
 
         if (oldCluster) {
-            oldCluster.nids = newCluster.nids;
-            oldCluster.addresses = newCluster.addresses;
```
Comment (Klaster1) [inline modules/web-console/frontend/app/modules/agent/AgentManager.service.ts]: @nva can you please explain this change? Did you change what properties are available on cluster object?

Comment (nva) [reply modules/web-console/frontend/app/modules/agent/AgentManager.service.ts]: I added more types and found this error. This changed when we rewrite to java backend and doesn't fixed by @vsisko

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/BrowsersService.java (hunk 6 above)
Comment (nva) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/BrowsersService.java]: No we just emulate same error in cluster

File: modules/web-console/frontend/app/modules/agent/types/ClusterSecrets.js (hunk 2 above)
Comment (nva) [reply modules/web-console/frontend/app/modules/agent/types/ClusterSecrets.js]: We still need user, password for case when session expired

File: modules/web-console/frontend/app/components/connected-clusters-dialog/components/list/column-defs.ts (hunk 1 above)
Comment (nva) [reply modules/web-console/frontend/app/components/connected-clusters-dialog/components/list/column-defs.ts]: All works fine.

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/repositories/IgniteSessionRepository.java (hunk 4 above)
Comment (nva) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/repositories/IgniteSessionRepository.java]: yes

File: modules/web-console/frontend/app/modules/agent/AgentManager.service.ts (hunk 8 above)
Comment (Klaster1) [reply modules/web-console/frontend/app/modules/agent/AgentManager.service.ts]: Good find! 👍

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/BrowsersService.java [hunk 9]
```diff
@@ -158,19 +188,55 @@ public void sendToAgent(AgentKey key, WebSocketRequest evt) {
...
                         throw new IllegalStateException(messages.getMessage("err.missing-cluster-id-param"));
 
-                    if (evt.getEventType().equals(NODE_VISOR))
-                        evt.setPayload(toJson(fillVisorGatewayTaskParams(payload)));
+                    TopologySnapshot top = clustersRepo.get(clusterId);
+
+                    if (top == null)
+                        throw new IllegalStateException(messages.getMessageWithArgs("err.cluster-not-found-by-id", clusterId));
+
+                    // TODO GG-24427 Append sessionId if it exists in session store.
```
[the remark this answers is not in the document]
Comment (nva) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/BrowsersService.java]: Removed

File: modules/web-console/web-console-server/src/test/java/org/apache/ignite/console/services/SessionsServiceTest.java (hunk 7 above)
Comment (nva) [reply modules/web-console/web-console-server/src/test/java/org/apache/ignite/console/services/SessionsServiceTest.java]: Added

**Comment by nva** [comment]
> * I approve more TS type coverage, but at the same time I am concerned about changes to method signatures where it seemingly has nothing to do with task at hand.
This just a types, they can't breaks nothing. All other changes related with task.

> * Please cover frontend changes with tests. How are you going to ensure that unrelated changes break nothing?
I don't have any idea how to cover this. Can you give me a example with secured cluster?

**Comment by Klaster1** [comment]
>I don't have any idea how to cover this. Can you give me a example with secured cluster?
I'd cover this with E2E tests and mock appropriate backend endpoints. You've changed some features unrelated to the feature, might be a good idea to cover that too in a same manner.

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/BrowsersService.java [hunk 10]
```diff
@@ -158,19 +188,54 @@ public void sendToAgent(AgentKey key, WebSocketRequest evt) {
...
+                    if (top == null)
+                        throw new IllegalStateException(messages.getMessageWithArgs("err.cluster-not-found-by-id", clusterId));
+
+                    if (top.isSecured()) {
+                        SessionAttribute sesAttr = new SessionAttribute(ses, SESSION_TOKEN_ATTR_PREFIX + clusterId);
+
+                        String sesTok = sesSrvc.get(sesAttr);
+
+                        if (F.isEmpty(sesTok) && Stream.of("user", "password").noneMatch(params::containsKey)) {
+                            sendMessageQuiet(ses, evt.response(RestResult.fail(STATUS_FAILED, "Failed to handle request - unknown session token (maybe expired session)")));
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/BrowsersService.java]: Why not to use EXPIRED_SESSION_ERROR_MESSAGE here?

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/BrowsersService.java [hunk 11]
```diff
@@ -319,54 +384,81 @@ protected void registerVisorTasks() {
...
+
+            try {
+                RestResult res = fromJson(evt0.getPayload(), RestResult.class);
+
+                switch (res.getSuccessStatus()) {
+                    case STATUS_SUCCESS:
+                        if (!F.isEmpty(res.getSessionToken()))
+                            sesSrvc.update(sesAttr, res.getSessionToken());
+
+                        break;
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/BrowsersService.java]: Add empty line after `break`;

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/BrowsersService.java [hunk 12]
```diff
@@ -319,54 +384,81 @@ protected void registerVisorTasks() {
...
+                    case STATUS_FAILED:
+                        if (res.getError().startsWith(EXPIRED_SESSION_ERROR_MESSAGE))
+                            sesSrvc.remove(sesAttr);
+
+                        break;
+                    default:
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/BrowsersService.java]: Add empty line after `break`;

Comment (nva) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/BrowsersService.java]: Added

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/BrowsersService.java (hunk 11 above)
Comment (nva) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/BrowsersService.java]: Added

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/BrowsersService.java (hunk 6 above)
Comment (nva) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/BrowsersService.java]: Use value from static variable

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/repositories/IgniteSessionRepository.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    private int maxInactiveInterval;
+
+    /** */
+    protected final TransactionManager txMgr;
+
+    /** Session cache. */
+    private IgniteCache<String, MapSession> cache;
+
+    /**
+     * @param ignite Ignite.
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/repositories/IgniteSessionRepository.java]: `@param txMgr` - is missing in javadocs

Comment (nva) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/repositories/IgniteSessionRepository.java]: Done

File: modules/management-console-agent/pom.xml [hunk 14]
```diff
@@ -51,19 +51,19 @@
         <dependency>
             <groupId>org.eclipse.jetty</groupId>
             <artifactId>jetty-client</artifactId>
-            <version>${jetty.version}</version>
+            <version>9.4.22.v20191022</version>
```
Comment (akuznetsov-gridgain) [inline modules/management-console-agent/pom.xml]: Why you changed version here? But not in parent/pom.xml ?

Comment (vsisko) [reply modules/management-console-agent/pom.xml]: It is discussed with @smoldachev in parent/pom.xml 9.4.18 version of jetty is returned.
9.4.22 version is used only for management-console-agent project.

Comment (vsisko) [reply modules/management-console-agent/pom.xml]: Version reverted.

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/common/SessionAttribute.java [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+        return sesId;
+    }
+
+    /**
+     * @return Attribute name.
+     */
+    public String getName() {
+        return name;
+    }
+}
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/common/SessionAttribute.java]: Missing toString()

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/TransitionService.java [hunk 16]
```diff
@@ -70,7 +70,7 @@
     /**
      * @param ignite Ignite.
      * @param agentsRepo Agents repository.
-     * @param clustersRepo Clusters repository.
+     * @param clustersRepo Repositories to work with clusters.
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/TransitionService.java]:
```suggestion
     * @param clustersRepo Repository to work with clusters.
```

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/common/SessionAttribute.java (hunk 15 above)
Comment (vsisko) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/common/SessionAttribute.java]: Added.

Comment (vsisko) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/common/SessionAttribute.java]: Added.

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/AgentsService.java [hunk 17]
```diff
@@ -265,16 +266,27 @@ public void revokeToken(Account acc, String oldTok) {
      * @param req Request.
      * @throws IllegalStateException If connected agent not founded.
      */
-    void sendLocally(AgentRequest req) throws IllegalStateException, IOException {
+    void sendLocally(AgentRequest req) throws IllegalStateException, IllegalArgumentException, IOException {
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/AgentsService.java]: Please update `@throws` section in javadocs.

Comment (vsisko) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/AgentsService.java]: Updated

**Pull request merged**